### PR TITLE
Allow setting every array element in replacements

### PR DESCRIPTION
**Pull request by koba1t**
Implement to https://github.com/kubernetes-sigs/kustomize/issues/4053.
And resolve https://github.com/kubernetes-sigs/kustomize/issues/4414#issuecomment-1018835598 and https://github.com/kubernetes-sigs/kustomize/issues/4391.

### CODE REVIEW

**Comment by natasha41575** [comment]
Would you mind also adding a test to https://github.com/kubernetes-sigs/kustomize/blob/master/api/filters/replacement/replacement_test.go to add some e2e coverage? Thank you for working on this

File: api/filters/replacement/replacement_test.go [hunk 1]
```diff
@@ -1431,6 +1431,136 @@ spec:
...
+        app: sample-deploy
+    spec:
+      containers:
+      - image: nginx
+        name: main
+        env:
+        - name: deployment-name
+          value: sample-deploy`,
+		},
+		"list index contains '*' character": {
```
Comment (natasha41575) [inline api/filters/replacement/replacement_test.go:1493]: what is the difference between this test and the above test?

Also, the test seems not to be passing.

Comment (koba1t) [reply api/filters/replacement/replacement_test.go:1493]: It tests to replace multi values.
- https://github.com/kubernetes-sigs/kustomize/pull/4424/files#diff-909c0819f13e268a60555f0aa78bd8acdbb42c10f3e2a906d62ba4b9d1722947R1515-R1522
- https://github.com/kubernetes-sigs/kustomize/pull/4424/files#diff-909c0819f13e268a60555f0aa78bd8acdbb42c10f3e2a906d62ba4b9d1722947R1556-R1563

**Comment by koba1t** [comment]
Sorry for the delay to reply and fix.
But, I found a problem with this code when writing test code.

I think "replacements" couldn't replace for the target to multi values.

e.g.

### source
```
# deploy.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: sample-deploy
  name: sample-deploy
spec:
  replicas: 1
  selector:
    matchLabels:
      app: sample-deploy
  template:
    metadata:
      labels:
        app: sample-deploy
    spec:
      containers:
      - image: nginx
        name: main
        env:
        - name: deployment-name
          value: XXXXX
        - name: foo
          value: bar
      - image: nginx
        name: sidecar
        env:
        - name: deployment-name
          value: YYYYY
```

```
# kustomization.yaml
resources:
  - deploy.yaml

replacements:
- source:
    kind: Deployment
    name: sample-deploy
    fieldPath: metadata.name
  targets:
  - select:
      kind: Deployment
    fieldPaths:
    - spec.template.spec.containers.[image=nginx].env.[name=deployment-name].value
```

### result
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: sample-deploy
  name: sample-deploy
spec:
  replicas: 1
  selector:
    matchLabels:
      app: sample-deploy
  template:
    metadata:
      labels:
        app: sample-deploy
    spec:
      containers:
      - env:
        - name: deployment-name
          value: sample-deploy
        - name: foo
          value: bar
        image: nginx
        name: main
      - env:
        - name: deployment-name
          value: YYYYY
        image: nginx
        name: sidecar
```

It's only to replace the first element in deploy.yaml.

I think it's needed to fix when using this function. And I'm trying it now.

**Comment by koba1t** [comment]
Hello @natasha41575. 
Every test case was passing. I think this work is complete.

Sorry for to delay.
If you have time, please re-check this PR.

**Comment by natasha41575** [comment]
> If you have time, please re-check this PR.

Thanks for the update! I have been out for a few days, will try to take a look soon.

File: api/filters/replacement/replacement.go [hunk 2]
```diff
@@ -119,20 +119,42 @@ func applyToNode(node *yaml.RNode, value *yaml.RNode, target *types.TargetSelect
 		if target.Options != nil && target.Options.Create {
 			t, err = node.Pipe(yaml.LookupCreate(value.YNode().Kind, fieldPath...))
 		} else {
-			t, err = node.Pipe(yaml.Lookup(fieldPath...))
+			// t, err = node.Pipe(yaml.Lookup(fieldPath...))
```
Comment (natasha41575) [inline api/filters/replacement/replacement.go:122]: Did you forget to remove this?

File: kyaml/yaml/fns.go [hunk 3]
```diff
@@ -796,6 +803,12 @@ func SplitIndexNameValue(p string) (string, string, error) {
 	return parts[0], parts[1], nil
 }
 
+// IsMatchEveryIndex returns true if p is matching every elements.
+// e.g. "*"
+func IsMatchEveryIndex(p string) bool {
```
Comment (natasha41575) [inline kyaml/yaml/fns.go:808]: Suggest changing the name here to `IsWildcard` and moving this up to line 791 to keep the `IsXXXX` functions together

**Review by natasha41575 — COMMENTED** [review]
Some very minor nits, apart from that LGTM.

File: api/filters/replacement/replacement.go (hunk 2 above)
Comment (koba1t) [reply api/filters/replacement/replacement.go]: Yes. I'll delete it.

**Comment by koba1t** [comment]
I fix any mistake from your review and add the above testcase(https://github.com/kubernetes-sigs/kustomize/pull/4424#issuecomment-1028230198).

**Comment by natasha41575** [comment]
/lgtm
/approve

Thanks!

**Comment by afirth** [comment]
you rock @koba1t

**Pull request merged**